File: UART_Utilities/UART.c
/*
    John Talbot
    July 25, 2016
    
    PiUART is a set of functions allowing the AVR and RasPi to 
    communicate over UART. For use with GardenPi code

    Code modified from USART.h and USART.c files from Make: Programming
    AVR by Elliot Williams for use with GardenPi package

    License included in UART_Utilities file
*/

#include <avr/io.h>
#include <util/delay.h>
#include <util/setbaud.h>
#include <avr/interrupt.h>
#include "UART.h"
#include "GardenPiUARTBytes.h"

//---------------------------------------------------------------
// INITIALIZATION
//---------------------------------------------------------------
/*
    This function will initialize UART communications on the AVR
    using speeds calculated by setbaud.h and using 8 data bits 
    with 1 stop bit
*/
void initUART(void) {
    // Values defined in setbaud.h
    UBRR0H = UBRRH_VALUE;
    UBRR0L = UBRRL_VALUE;

    // Use double speed Tx/Rx if flag is set in setbaud.h
#if USE_2X
    UCSR0A |= (1 << U2X0);
#else
    UCSR0A &= ~(1 << U2X0);
#endif
  
    // Enable UART transmitter/receiver
    UCSR0B = (1 << TXEN0) | (1 << RXEN0);

    // Enable UART receive interrupt in register if flag enabled
    // in UART.h header file
    if(ENABLE_RX_INTERRUPT){
        UCSR0B |= (1 << RXCIE0);
    }
    else{
        UCSR0B &= ~(1 << RXCIE0);
    }

    // 8 data bits, 1 stop bit
    UCSR0C = (1 << UCSZ01) | (1 << UCSZ00);
}

//--------------------------------------------------------------
// LOW LEVEL Tx/Rx FUNCTIONS
//--------------------------------------------------------------
//This function will transmit one byte of data over UART
void transmitByte(uint8_t data) {
    // Loop until transmit data register is empty
    while(!(UCSR0A & (1 << UDRE0))){}

    // Write byte to transmit data register to send
    UDR0 = data;
}

// This function will receive one byte of data over UART
uint8_t receiveByte(void) {
    // Loop until receive data register has data
    while(!(UCSR0A & (1 << RXC0))){}

    // Read data from receive register
    return UDR0;
}

//--------------------------------------------------------------
// MID LEVEL MESSAGE FUNCTIONS
//--------------------------------------------------------------
// Handle UART communication error
void commError(void){
    transmitByte(UART_COMM_ERROR);
    transmitByte(END_MSG);

    // Flush receive buffer by temporarily disabling RXEN0
    UCSR0B &= ~(1 << RXEN0);
    _delay_us(25);
    UCSR0B |= (1 << RXEN0);
}

void transmitReady(void){
    transmitByte(AVR_READY);
    transmitByte(END_MSG);
}

void transmitRequest(void){
    transmitByte(AVR_REQ_RASPI);
    transmitByte(END_MSG);
}

void transmitConfirm(void){
    transmitByte(AVR_REC_MSG_CONFIRM);
    transmitByte(END_MSG);
}

void transmitInitialize(void){
    transmitByte(AVR_INIT_TO_RASPI);
    transmitByte(END_MSG);
}

//--------------------------------------------------------------
// HIGH LEVEL MESSAGE FUNCTIONS
//--------------------------------------------------------------
// This function will receive a full request message from a
// RasPi and return proper confirmation bytes
int * receiveMessage(void){
    // Hold inital byte received
    uint8_t i=0;
    uint8_t j=0;
    uint8_t n=0;

    // Array to hold received data
    static int msgArray[10];
    int dataByteIn[2];

    // Receive initial byte
    for(j=0; j<2; j++){
        dataByteIn[j] = receiveByte();
    }

    // If message is not properly terminated write error code
        // to entire array and call commError()
    if(dataByteIn[1] != END_MSG){
        for (i=0; i<4; i++){
            msgArray[i] = UART_COMM_ERROR;
        }
        commError();
    }

    // RasPi normal request message to AVR
    else if (dataByteIn[0] == RASPI_REQ_AVR){
        // Transmit AVR ready to receive message
        transmitReady();
        
        // Loop through until EOM reached and store in array
        do{
            msgArray[n] = receiveByte();
            n++;
        }while(msgArray[n-1] != END_MSG);
        transmitConfirm();
    }

    // RasPi initialize request to AVR
    else if(dataByteIn[0] == RASPI_INIT_TO_AVR){
	    // Transmit AVR initialized ready byte
        transmitInitialize();

        // Write initialized byte to array
        for(i=0; i<4; i++){
            msgArray[i] = AVR_INIT_TO_RASPI;
        }
    }

    // Improper initial communication byte, call commError()
    else{
        commError();
    }

    return msgArray;
}

// This function will transmit a full message to RasPi
uint8_t transmitMessage(volatile int sendData[4]){
    // Initialize loop counter
    uint8_t i;
    uint8_t badSendFlag = 0;

    // Clear global interrupt to stop receive interrupt
    cli();

    // Send AVR request RasPi byte
    transmitRequest();

    // Wait for response from RasPi
    // **** This has possibility of hanging program -- needs improvement!
    if (receiveByte() == RASPI_READY){
        // Remove EOM from buffer
        receiveByte();

        // Loop to send all of data array
        for(i=0; i<4; i++){
            transmitByte(sendData[i]);
        }

        // Transmit end of message byte
        transmitByte(END_MSG);

        // Check for good confirm byte
        if (receiveByte() == RASPI_REC_MSG_CONFIRM){
            // Remove EOM from buffer
            receiveByte();
        }
        else{
            // Set bad send flag
            badSendFlag = 2;
        }
    }

    // If improper response is received, call commError()
    else{
        // Set bad send flag
            badSendFlag = 1;
    }

    // If badSendFlag is set - call commError()
    if(badSendFlag != 0){
        commError();
    }

    // reenable global interrupts
    sei();

    // Return value of badSendFlag
    return badSendFlag;
}

File: UART.py
#------------------------------------------------------------------
# HEADER
#------------------------------------------------------------------
# John Talbot
# July 26, 2016
# This module will define functions for use in the GardenPi project

#------------------------------------------------------------------
# IMPORT MODULES
#------------------------------------------------------------------
import UART_Messages
import time
import serial
import struct
import os

#------------------------------------------------------------------
# SERIAL PORT PARAMETERS
#------------------------------------------------------------------
SERIAL_PORT = "/dev/ttyAMA0"
BAUD_RATE = 9600
DEF_TIMEOUT = 30
PACKAGE_LEN = 4
sp = 0

#------------------------------------------------------------------
# INITIALIZATION FUNCTION
#------------------------------------------------------------------
def initUART(port=SERIAL_PORT, baud=BAUD_RATE, timeOutVar=DEF_TIMEOUT):
	# Echo message to user
	print("Initializing UART Communications... Please Wait")
	time.sleep(3)

	# Open serial port and map to global variable sp
	global sp
	sp = serial.Serial(port, baud, timeout=timeOutVar)
	sp.flushInput()
	sp.flushOutput()

	# Send initialization byte to AVR
	transmitInitial()

	# Wait for handshake byte from AVR
	handshakeByte = sp.readline()[:-1]

	# Errorproof initialization
	if (handshakeByte != UART_Messages.AVR_INIT_TO_RASPI):
		os.system('clear')
		print(handshakeByte)
		raise Exception("Invalid 'AVR Initialization' return")
	else:
		os.system('clear')
		print("UART Communications Initialized!")
		time.sleep(3)
		os.system('clear')

#------------------------------------------------------------------
# LOW LEVEL UART FUNCTIONS
#------------------------------------------------------------------
# Reply to request with ready byte
def transmitReady():
	sp.write(UART_Messages.RASPI_READY)
	sp.write(UART_Messages.END_MSG)

# Send request to AVR
def transmitRequest():
	sp.write(UART_Messages.RASPI_REQ_AVR)
	sp.write(UART_Messages.END_MSG)

# Send confirm byte to AVR
def transmitConfirm():
	sp.write(UART_Messages.RASPI_REC_MSG_CONFIRM)
	sp.write(UART_Messages.END_MSG)

# Send initialization byte to AVR
def transmitInitial():
	sp.write(UART_Messages.RASPI_INIT_TO_AVR)
	sp.write(UART_Messages.END_MSG)

# Read and check AVR ready byte
def readReady():
	# Wait for and check AVR return ready byte
	readyByte = sp.readline()[:-1]
	if (readyByte != UART_Messages.AVR_READY):
		raise Exception("Invalid 'AVR Ready' return")
	else:
		print("AVR Ready")
		time.sleep(2)
		os.system('clear')

# Handle AVR confirm message
def readConfirm():
	# Wait for and check AVR return message received byte
	confirmByte = sp.readline()[:-1]
	if (confirmByte != UART_Messages.AVR_REC_MSG_CONFIRM):
		raise Exception("Invalid 'AVR Message Confirm' return")
	else:
		print("AVR confirmed message receipt")

#------------------------------------------------------------------
# MID LEVEL UART FUNCTIONS
#------------------------------------------------------------------
def transmitPackage( dataPackage ):
	# Check for empty data package
	if (len(dataPackage) < 1):
		raise Exception("Empty Data Package")
	else:
		# Initialize counter and length
		i = 0
		length = len(dataPackage)

		# Loop to send all bytes in data package
		while(i < length):
			sp.write(dataPackage[i])
			i += 1;

		# Send EOM byte
		sp.write(UART_Messages.END_MSG)

# Read in package, strip EOM character, and encode into hex
def readPackage():
	# Initialize array to return data bytes
	dataOut = [];

	# Read in package and strip newline
	packageIn = sp.readline()[:-1]

	# Check package length - If != 4 throw exception
	packageLength = len(packageIn)
	if (packageLength != PACKAGE_LEN):
		raise Exception("Incorrect Package Length")
	else:
		dataOut.append(bytes([packageIn[0]]))
		dataOut.append(bytes([packageIn[1]]))
		dataOut.append(convertData(packageIn[2:4]))
		return dataOut

# Check data message for errors - If none convert to int
def convertData(dataIn):
	# Data message should be 2 bytes long
	msgLen = len(dataIn)
	if (msgLen != 2):
		raise Exception("Incorrect Data Length for convertData()")
	elif (bytes([dataIn[0]]) >= UART_Messages.UART_COMM_ERROR):
		parseError(dataIn[0])
	else:
		dataRet = struct.unpack('>H',dataIn)
		return dataRet

# Parse error code
def parseError(errorCode):
	errHex = bytes([errorCode])
	if (errHex == UART_Messages.UART_COMM_ERROR):
		raise Exception("Communications Error with AVR")
	elif (errHex == UART_Messages.HCSR04_TIMEOUT):
		raise Exception("HC-SR04 Sensor Timeout Error")
	else:
		raise Exception("Unrecognized Error Code")

#------------------------------------------------------------------
# HIGH LEVEL UART FUNCTIONS
#------------------------------------------------------------------
def transmitMessage( dataMessage ):
	# Send AVR request byte
	transmitRequest()

	# Wait for AVR ready
	readReady()

	# Send data package
	transmitPackage( dataMessage )

	# Wait for AVR confirm
	readConfirm()
	



